src/services/quotes/quote-sources/sovryn-source.ts

```diff
@@ -0,0 +1,82 @@
+import { Chains } from '@chains';
+import { Chain, Address } from '@types';
+import { Addresses } from '@shared/constants';
+import { IQuoteSource, QuoteParams, QuoteSourceMetadata, QuoteSourceSupport, SourceQuoteRequest, SourceQuoteResponse } from './types';
+import { addQuoteSlippage, failed } from './utils';
+import { isSameAddress } from '@shared/utils';
+import { DEFAULT_SWAP_ROUTES, SmartRouter, smartRoutes } from '@sovryn/sdk';
+import { AlwaysValidConfigAndContextSource } from './base/always-valid-source';
+
+export const SOVRYN_METADATA: QuoteSourceMetadata<SovrynSupport> = {
+  name: 'Sovryn',
+  supports: {
+    chains: [Chains.ROOTSTOCK.chainId],
+    swapAndTransfer: false,
+    buyOrders: false,
+  },
+  logoURI: 'ipfs://QmUpdb1zxtB2kUSjR1Qs1QMFPsSeZNkL21fMzGUfdjkXQA',
+};
+
+type SovrynSupport = { buyOrders: false; swapAndTransfer: false };
+export class SovrynQuoteSource extends AlwaysValidConfigAndContextSource<SovrynSupport> {
+  getMetadata() {
+    return SOVRYN_METADATA;
+  }
+
+  async quote({
+    components: { providerService },
+    request: {
+      chain,
+      sellToken,
+      buyToken,
+      order,
+      accounts: { takeFrom },
+      config: { slippagePercentage },
+    },
+  }: QuoteParams<SovrynSupport>): Promise<SourceQuoteResponse> {
+    // Get Rootstock provider
+    const provider = providerService.getEthersProvider({ chainId: Chains.ROOTSTOCK.chainId });
+    // Create smart router
+    // FIX: Enable routes ZeroRedemption, MoC and Mynt once they enable it's use without permit.
+    const smartRouter = new SmartRouter(provider, Object.values([smartRoutes.ammSwapRoute]));
+    // Map buy and sell token since on Rootstock native token = address(0)
+    const mappedSellToken = mapNativeToken(sellToken);
+    const mappedBuyToken = mapNativeToken(buyToken);
+    // Find best quote
+    try {
+      const result = await smartRouter.getBestQuote(mappedSellToken, mappedBuyToken, order.sellAmount);
+      // Get swap tx data
+      const swapTxData = await result.route.swap(mappedSellToken, mappedBuyToken, order.sellAmount, takeFrom, {
+        slippage: slippagePercentage * 100,
+      });
+
+      if (!swapTxData || !swapTxData.to) {
+        throw new Error('Failed to calculate a quote');
```

Comment:
```suggestion
        failed(SOVRYN_METADATA, chain, sellToken, buyToken, 'Failed to calculate quote');
```

Reply:
I had this in place, but if I through it, it will do the same through the catch clause. wdyt?

Reply:
Good point, feel free to keep it as is